docs/notebooks/doc2vecc_newsgroup20.ipynb

```diff
@@ -0,0 +1,372 @@
+{
+ "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "import sklearn\n",
+    "import gensim\n",
+    "import logging\n",
+    "\n",
+    "import numpy as np\n",
+    "import pandas as pd\n",
+    "\n",
+    "logging.basicConfig(format='%(asctime)s : %(levelname)s : %(message)s', level=logging.INFO)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 122,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "newsgroup = sklearn.datasets.fetch_20newsgroups(subset=\"all\", shuffle=True, random_state=42, \n",
+    "                                                remove=('header', 'footer', 'quotes'))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 139,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "processed_docs = [gensim.utils.simple_preprocess(x) for x in newsgroup['data']]\n",
+    "\n",
+    "X_train, X_test, _, y_test = sklearn.model_selection.train_test_split(processed_docs, newsgroup['target'], \n",
+    "                                                                      random_state=42, test_size=0.33)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "doc2vecc_model = gensim.models.word2vec.Word2Vec(X_train, size=100, alpha=0.025, window=5, min_count=3, \n",
+    "                                                 sample=1e-3, seed=42, workers=4, min_alpha=0.0001, \n",
+    "                                                 sg=1, hs=0, negative=5, ns_exponent=0.75, cbow_mean=0, \n",
+    "                                                 iter=30, doc2vecc=0.2)\n",
+    "\n",
+    "# 2018-11-27 13:57:23,705 : INFO : training on a 80221560 raw words (64035464 effective words) took 4313.5s, 14845 effective words/s"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "word2vec_model = gensim.models.word2vec.Word2Vec(X_train, size=100, alpha=0.025, window=5, min_count=3, \n",
+    "                                                 sample=1e-3, seed=42, workers=4, min_alpha=0.0001, \n",
+    "                                                 sg=1, hs=0, negative=5, ns_exponent=0.75, cbow_mean=0, \n",
+    "                                                 iter=30)\n",
+    "\n",
+    "# 2018-11-27 14:05:27,024 : INFO : training on a 80221560 raw words (64033783 effective words) took 252.7s, 253358 effective words/s"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 141,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [],
+   "source": [
+    "def infer_vectors(model, documents):\n",
```

Comment:
I'd avoid calling this `infer_vector()`, since it's just using a simple average, rather than an iterative 'inference' process (like the same-named `Doc2Vec` method). Also: an 'average_vector(str[])` utility method probably belongs on `Word2Vec` or one of its parent interfaces, since it's such a common need. It might already exist or be factorable-out-of one of the existing methods that already averages-several-words-together. It could potentially include optional parameters to limit the considered words to those above a certain frequency or in the top-N of all words, or toggle on or off vector norming before or after averaging.

Reply:
The KeyedVectors functions `most_similar()` and `doesnt_match()` both internally compute a mean-of-multiple-vectors, in slightly different ways. (In `most_similar()`, some of the `negative` contributors to the mean are negated before mixed-in. In both, numpy functions offer compactness & efficiency that might be better (due to increased use of native parallel array operations) than an explicit python accumulate-loop-then-divide.